godot-ffi/src/godot_ffi.rs

```diff
@@ -281,37 +368,23 @@ mod scalars {
             impl GodotFuncMarshal for $T {
                 type Via = $Via;
 
-                unsafe fn try_from_sys(ptr: sys::GDExtensionTypePtr) -> Result<Self, $Via> {
-                    let via = <$Via as GodotFfi>::from_sys(ptr);
-                    Self::try_from(via).map_err(|_| via)
-                }
-
-                unsafe fn try_write_sys(&self, dst: sys::GDExtensionTypePtr) -> Result<(), Self> {
-                    <$Via>::try_from(*self)
-                        .and_then(|via| {
-                            <$Via as GodotFfi>::write_sys(&via, dst);
-                            Ok(())
-                        })
-                        .map_err(|_| *self)
-                }
-            }
-        };
-
-        ($T:ty as $Via:ty; lossy) => {
-            // implicit bounds:
-            //    T: TryFrom<Via>, Copy
-            //    Via: TryFrom<T>, GodotFfi
-            impl GodotFuncMarshal for $T {
-                type Via = $Via;
-
-                unsafe fn try_from_sys(ptr: sys::GDExtensionTypePtr) -> Result<Self, $Via> {
-                    let via = <$Via as GodotFfi>::from_sys(ptr);
+                unsafe fn try_from_arg(
+                    ptr: sys::GDExtensionTypePtr,
+                    call_type: CallType,
+                ) -> Result<Self, $Via> {
+                    let via = <$Via as GodotFfi>::from_arg_ptr(ptr, call_type);
+                    // Godot does not always give us a valid `i64`/`f64`, nor `Self` when it does a pointer
+                    // call, so we cannot use `try_from` here.
                     Ok(via as Self)
                 }
```

Comment:
ill look into it again i just remember that we'd get weird inconsistent values when going through the `Via` type because we'd be interpreting garbage bits as actual bits in the integer or something.

Reply:
But then that would be another bug. In any case, "garbage values" wouldn't just be fixed by truncating integers. At best, you hide some problems.

In general, we must make sure we use the same type on Rust side as we do on C (Godot API) side. Not every integer type can be passed to Godot, they typically use `i64`.